doc/HLD_of_SNS_Client.rst

```diff
@@ -0,0 +1,283 @@
+
+ Conformance
+=============
+
+- 1[u.FOP] ST
+
+- 2[u.LAYOUT.PARAMETRIZED] ST 
+
+- 3[u.FOP.SNS] ST 
+
+- 4[u.LAYOUT.DATA] ST 
+
+- 5[u.LAYOUT.POOLS] ST 
+
+- 6[u.lib.allocate-page] 
+
+- 7[u.fop.cache.add] 
+
+- 8[u.fop.cache.unplug] 
+
+- 9[u.fop.rpc.to] 
+
+- 10[u.FOP.BATCHING] ST
+
+- [r.sns-client.nocache]: holds per caching policy described in the rpc sub-section.
+
+- [r.sns-client.norecovery]: holds obviously;
+
+- [R.M0.IO.DIRECT]: no-caching and 0-copy for data together implement direct-io.
+
+ Dependencies
+===============
+
+- layout:
+
+  - [u.LAYOUT.SNS] ST: server network striping can be expressed as a layout
+
+  - [u.LAYOUT.DATA] ST: layouts for data are supported
+
+  - [u.LAYOUT.POOLS] ST: layouts use server and device pools for object allocation, location, and identification
+
+  - [u.LAYOUT.PARAMETRIZED] ST: layouts have parameters that are substituted to perform actual mapping
+
+- fop:
+
+  - [u.fop] ST: M0 uses File Operation Packets (FOPs)
+
+  - [u.fop.rpc.to]: a fop can be serialized into an rpc
+
+  - [u.fop.nrs] ST: FOPs can be used by NRS
+
+  - [u.fop.sns] ST: FOP supports SNS
+
+  - [u.fop.batching] ST: FOPs can be batched in a batch-FOP
+
+  - [u.fop.cache.put]: fops can be cached
+
+  - [u.fop.cache.unplug]: fop cache can be de-staged forcibly
+
+- NRS:
+
+  - [u.nrs] ST: Network Request Scheduler optimizes processing order globally
+
+- misc:
+
+  - [u.io.sns] ST: server network striping is supported
+ 
+  - [u.lib.allocate-page]: page allocation interface is present in the library
+  
+  Security Model
+================
+ 
+ Security is outside of scope of the present design.
+ 
+ Refinement
+============
+ 
+ Detailed level design specification should address the following:
+ 
+ - concurrency control and liveness rules for fops and layouts;
+
+ - data structures for mapping between layout and target objects in the pool;
+
+ - instantiation of a layout formula;
+
+ - relationship between fop and its sub-fops: concurrency control, liveness, ownership
+ 
+********
+ State
+********
+
+State diagrams are part of the detailed level design specification.
+
+***********
+ Use Cases
+***********
+
+ Scenarios
+===========
+
+Scenario 1: [usecase.sns-client-read]
+
+Relevant quality attributes: usability
+
+Stimulus: an incoming read operation request from a user space operation
+
+Stimulus source: a user space application, potentially meditated by a loop-back device driver
+
+Environment: normal client operation
+
+Artifact: call to VFS ->read() entry point
+
+Response: a fop is created, network transmission of operation parameters to all involved data servers is started as specified by the file layout, servers place retrieved data directly in user buffers, once transmission completes, the fop is destroyed.
+
+Response measure: no data copying in the process
+
+Questions and issues:
+
```

Comment:
Can you please convert this into a table?
Please refer to the previous PR I have done.

Reply:
Incorporated.